### PR TITLE
[jit][script] enable advanced indexing with tensors

### PR DESCRIPTION
**Pull request by zou3519**
On the way to #10774

This PR adds advanced indexing with tensors.
The approach is to desugar advanced indexing into an at::index op.
This is exactly how normal pytorch does it.
[(I used this code as reference)](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp)

Supporting sequences is a little tricky because JIT script doesn't have
an easy way to turn arbitrary n-dimensional python lists into a tensor
(it would be easy if we supported `torch.tensor`), so that'll come
in a future PR.

cc @jamesr66a @zdevito

### CODE REVIEW

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -1616,6 +1625,16 @@ struct to_ir {
...
     }
     size_t dim = 0;
+
+    // Accumulate tensor Values here; call aten::index on them later.
+    std::vector<Value*> tensor_indices;
+
+    auto handle_tensor = [&](Value* tensor) {
+      tensor_indices.resize(dim + 1);
+      tensor_indices[dim] = tensor;
+      dim++;
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: How is that different than `tensor_indices.push_back(tensor); dim++`? If the vector can have holes can you add a comment saying how exactly those indices should be understood? Do they correspond to indexing before or after processing the slices?

Comment (apaszke) [reply torch/csrc/jit/script/compiler.cpp]: Also, it seems to be used in a single place only, so it shouldn't be necessary.

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -1668,11 +1704,18 @@ struct to_ir {
...
+      auto * index = idx.value(*graph);
+      const auto index_type = index->type();
+      if (index_type->isSubtypeOf(DynamicType::get())) {
+        return emitIndex(loc, gatherable.value(*graph), {index});
+      } else if (index_type == IntType::get()) {
+        return emitSelect(
+            loc, gatherable.value(*graph), /*dim=*/0, index);
+      }
+      throw ErrorReport(loc)
+        << "Cannot index tensor with value of type " << index_type->str();
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: It's really annoying that we have this thing implemented in multiple places. Can't we just call `emitMultidimIndexing` in here?

File: torch/csrc/jit/script/compiler.cpp (hunk 1 above)
Comment (zou3519) [reply torch/csrc/jit/script/compiler.cpp]: Yes, the vector can have holes in it. I'll add a comment saying how those are understood (`at::index` requires holes at indexes that aren't being indexed). 

handle_tensor is used in one place but in the future it'll be used in more places. Supporting indexing with sequences can be implemented by converting the sequence to a tensor and then using `at::index` on it.

File: torch/csrc/jit/script/compiler.cpp (hunk 2 above)
Comment (zou3519) [reply torch/csrc/jit/script/compiler.cpp]: Can do. The code is a little annoying to work around because emitMultidimIndex assumes that only the `gatherable` has been emitted, but `emitIndex` emits both the `gatherable` and the `index` and I would have to tell `emitIndex` to only emit the `gatherable`. I'll do some refactoring.

**Review by zdevito — APPROVED** [review]
This looks good to me.
